events/handler.go

```diff
@@ -0,0 +1,180 @@
+	LastNEvents int
+}
+
+// EventCategory is an enumerated type which lists the categories under which
+// events may fall. The Event field EventType is populated by this enum.
+type EventCategory int
+
+const (
+	OomInstanceCat EventCategory = iota
+	ContainerCreation
+	ContainerDeletion
+)
+
+// a general interface which populates the Event field EventData. The actual
+// object, such as an OomInstance, is set as an Event's EventData
+type EventDataInterface interface {
+}
+
+// returns a pointer to an initialized Events object
+func NewEventManager() *events {
+	return &events{
+		eventlist: make(EventSlice, 0),
+	}
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+	return &Request{
+		EventType: map[EventCategory]bool{},
+	}
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Len() int {
+	return len(e)
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Less(i, j int) bool {
+	return e[i].TimeOfEvent.Before(e[j].TimeOfEvent)
+}
+
+// sorts and returns up to the last N chronological elements
+func getLastNEvents(request *Request, eSlice EventSlice) EventSlice {
+	sort.Sort(eSlice)
+	n := request.LastNEvents
+	if n >= eSlice.Len() || n <= 0 {
+		return eSlice
+	}
+	return eSlice[eSlice.Len()-n:]
+}
+
+// determines if an event occurs within the time set in the request object and is the right type
+func checkIfEventSatisfiesRequest(request *Request, event *Event) bool {
+	startTime := request.StartTime
+	endTime := request.EndTime
+	eventTime := event.TimeOfEvent
+	if !startTime.IsZero() {
+		if startTime.After(eventTime) {
+			glog.Infof("start time ins't o")
+			return false
+		}
+	}
+	if !endTime.IsZero() {
+		if endTime.Before(eventTime) {
+			glog.Infof("end time isn't zero")
+			return false
+		}
+	}
+	if request.EventType[event.EventType] != true {
+		glog.Infof("event type is not found")
+		return false
+	}
+	return true
+}
+
+// method of Events object that screens Event objects found in the eventlist
+// attribute and if they fit the parameters passed by the Request object,
+// adds it to a slice of *Event objects that is returned
+func (self *events) GetEvents(request *Request) (EventSlice, error) {
+	returnEventList := EventSlice{}
+	for _, e := range self.eventlist {
+		if checkIfEventSatisfiesRequest(request, e) {
+			returnEventList = append(returnEventList, e)
+		}
+	}
+	returnEventList = getLastNEvents(request, returnEventList)
+	return returnEventList, nil
+}
+
+// method of Events object that maintains an *Event channel passed by the user.
+// When an event occurs that satisfies the parameters in the passed Request
+// object it is fed to the channel
+func (self *events) WatchEvents(eventChannel chan *Event) error {
```

Comment:
I agree that 4 is the most likely way to go, but I also realized that you refer to WatchEvents() as a way to take events in, it should be a way to get events out. GetEvents() lets you query historical data, while WatchEvents() lets you wait for new events that will occur. The only way to create events are through AddEvent().

The manager.Manager will have an `eventManager` as a member of its struct. On startup the manager.Manager will watchForOoms() and new/deleted containers. When either of those are encountered, it calls eventManager.AddEvent(). I tried adding more detail to the EventManager interface you have to see if that make it clearer.

```Go
type EventManager interface {
	// Report the occurrence of an event. This event is not available to GetEvents() 
	// requests. If there are any outstanding WatchEvents() looking for this type of event,
	// this new event will be delivered to them.
	AddtEvent(event Event) error

	// Allows the called to wait for events to occur. The caller with use the outChannel to 
	// receive any new events of the specified types.
	WatchEvents(outChannel chan Event, types map[EventCategory]struct{}) error

	// Allows the caller to fetch historical events. The request argument allows the called to 
	// filter the events returned.
	GetEvents(request Request) []Event
}
```